src/optimizers/optimizer.cpp

```diff
@@ -201,6 +201,48 @@ void optimizer::clear_gradient() {
   m_gradient_sources.clear();
 }
 
+AbsDistMat& optimizer::get_gradient_buffer(DataType& buf_scale,
+                                           DataType& in_scale,
+                                           bool allreduce_needed) {
+  if (m_gradient == nullptr) {
+    LBANN_ERROR("attempted to access gradient before it is set up");
+  }
+
+  // Complete outstanding allreduce.
+  if (m_gradient_status == optimizer_gradient_status::allreduce_started) {
+    finish_gradient_allreduce();
+  }
+  // Determine scaling factor and transition state.
+  switch (m_gradient_status) {
+  case optimizer_gradient_status::ready:
+    buf_scale = DataType(1);
+    in_scale = DataType(1);
+    if (allreduce_needed) {
+      buf_scale /= m_gradient->RedundantSize();
```

Comment:
Is redundant size here equal to the mini-batch size or the effective-mini-batch size?

Reply:
Neither. This is needed to properly combine contributions that will not (or have already been) allreduced with contributions that will be allreduced. Compare with `add_to_gradient`; this is doing precisely the same computation, just changing who is responsible for different aspects.